v2/internal/controllers/to_azure_secrets_test.go

```diff
@@ -79,10 +80,118 @@ func Test_MissingSecretKey_ReturnsError(t *testing.T) {
+	secretName := tc.Namer.GenerateName("admin-pw")
+	nameSpaceName := "secret-namespace"
+
+	createNamespaces(tc, nameSpaceName)
+	secretInDiffNamespace := createNamespacedSecret(tc, nameSpaceName, secretName)
+
+	rg := tc.CreateTestResourceGroupAndWait()
+
+	vnet := newVMVirtualNetwork(tc, testcommon.AsOwner(rg))
+	subnet := newVMSubnet(tc, testcommon.AsOwner(vnet))
+	networkInterface := newVMNetworkInterface(tc, testcommon.AsOwner(rg), subnet)
+
+	tc.CreateResourceAndWait(vnet)
+	tc.CreateResourceAndWait(subnet)
+	tc.CreateResourceAndWait(networkInterface)
+
+	vm := newVirtualMachine20201201(tc, rg, networkInterface, secretInDiffNamespace)
+
+	tc.CreateResourceAndWaitForState(vm, metav1.ConditionFalse, conditions.ConditionSeverityWarning)
+	// We expect the ready condition to include details of the error
+	tc.Expect(vm.Status.Conditions[0].Reason).To(Equal(conditions.ReasonSecretNotFound))
+	tc.Expect(vm.Status.Conditions[0].Message).To(
+		ContainSubstring("failed resolving secret references: %s/%s does not exist", tc.Namespace, secretInDiffNamespace.Name))
+
+	// Delete VM and resources.
+	tc.DeleteResourcesAndWait(vm, networkInterface, subnet, vnet, rg)
+}
+
+func Test_UserSecretInDifferentNamespace_ShouldNotTriggerReconcile(t *testing.T) {
+	t.Parallel()
+	ns1 := "secret-in-different-namespace-ns-1"
+	ns2 := "secret-in-different-namespace-ns-2"
+
+	cfg, err := testcommon.ReadFromEnvironmentForTest()
+	if err != nil {
+		t.Fatal(err)
+	}
+	cfg.TargetNamespaces = []string{ns1, ns2}
+	tc := globalTestContext.ForTestWithConfig(t, cfg)
+
+	createNamespaces(tc, ns1, ns2)
+
+	secretName := tc.Namer.GenerateName("server-secret")
+	ns1Secret := createNamespacedSecret(tc, ns1, secretName)
+	ns2Secret := createNamespacedSecret(tc, ns2, secretName)
+
+	rg := tc.NewTestResourceGroupWithNamespace(ns1)
+	tc.CreateResourceGroupAndWait(rg)
+
+	rg2 := tc.NewTestResourceGroupWithNamespace(ns2)
+	tc.CreateResourceGroupAndWait(rg2)
+
+	server1, _ := newFlexibleServer(tc, rg, ns1Secret)
+	server1.ObjectMeta = metav1.ObjectMeta{
+		Name:      tc.Namer.GenerateName("mysql"),
+		Namespace: ns1,
+	}
+	tc.CreateResourceAndWait(server1)
+
+	resourceID := genruntime.GetResourceIDOrDefault(server1)
+	tc.Expect(resourceID).ToNot(BeEmpty())
+
+	// Deleting server1 here using AzureClient so that Operator does not know about the deletion
+	_, err = tc.AzureClient.DeleteByID(tc.Ctx, resourceID, server1.GetAPIVersion())
+	if err != nil {
+		return
+	}
+
+	server2, _ := newFlexibleServer(tc, rg2, ns2Secret)
+	server2.ObjectMeta = metav1.ObjectMeta{
+		Name:      tc.Namer.GenerateName("mysql2"),
+		Namespace: ns2,
+	}
+
+	tc.CreateResourceAndWait(server2)
+
+	// Here we want to make sure that the server1 we deleted from Azure was not created again by a reconcile triggered when the second secret was created
+	// in other namespace
+	_, err = tc.AzureClient.GetByID(tc.Ctx, resourceID, server1.GetAPIVersion(), server1)
+	tc.Expect(err).ToNot(BeNil())
```

Comment:
Test for a specific `NotFound` error?

Reply:
Also consider issuing a fresh GET on `server1` in k8s and checking that its Ready condition is the same as before as well. Otherwise there could be a race where the operator is in the process of creating that new server but it's not yet done creating in Azure so this GET fails but really it did reconcile. If you also check Ready condition and that condition is _still_ Ready (and not `Reconciling` or something like that, then you know we're good.

Reply:
Done!